app/src/main/java/com/hedvig/app/feature/offer/ui/checkout/CheckoutActivity.kt

```diff
@@ -25,6 +34,58 @@ class CheckoutActivity : BaseActivity(R.layout.activity_checkout) {
             subtitle.text = parameter.subtitle
             val link = getString(R.string.OFFER_FOOTER_GDPR_INFO, parameter.gdprUrl)
             text.setMarkdownText(link)
+
+            emailEditText.addTextChangedListener {
+                viewModel.onEmailChanged(
+                    input = it?.toString() ?: "",
+                    hasError = emailInputContainer.error != null
+                )
+            }
+            emailEditText.setOnFocusChangeListener { _, hasFocus ->
+                if (!hasFocus) {
+                    viewModel.validateInput()
+                }
+            }
+            identityNumberEditText.addTextChangedListener {
+                viewModel.onIdentityNumberChanged(
+                    input = it?.toString() ?: "",
+                    hasError = identityNumberInputContainer.error != null
+                )
+            }
+            identityNumberEditText.setOnFocusChangeListener { _, hasFocus ->
+                if (!hasFocus) {
+                    viewModel.validateInput()
+                }
+            }
+
+            signButton.setOnClickListener {
+                viewModel.validateInput()
+            }
+        }
+
+        lifecycleScope.launch {
+            viewModel.viewState
+                .flowWithLifecycle(lifecycle)
+                .collect(::setInputState)
+        }
```

Comment:
I've tried a few different forms for expressing this code, and my personal preference has become the following:

```suggestion
        viewModel
        	.viewState
            .flowWithLifecycle(lifecycle)
            .onEach(::setInputState)
            .launchIn(lifecycleScope)
```

for the simple reason that it creates one less indentation 😅

What do you think?

Reply:
Very nice, totally agree.